### ISSUE TITLE
Deadlock during nng_close() - multi platform

### ISSUE THREAD

**Issue by mikisch81** [nanomsg/nng#1813]
**Describe the bug**

Continued from the closed https://github.com/nanomsg/nng/issues/1543

When calling `nng_close()` there is sometimes a deadlock which causes `nng_close()` to hang.
This happens also when using only sync APIs (no AIOs).

**Expected behavior**

`nng_close` should finish successfully.

**Actual Behavior**

`nng_close()` hangs.

**To Reproduce**

I created a modified version of the `reqrep` example code here (I use it with IPC transport): [https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3](https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3)

In the modified example in the client code right before calling `nng_recv` for the reply from the server I start a thread which just calls `nng_close`, after a couple of successful runs the deadlock happen:

```
mischwar@tlv-mpawy reqrep % ./reqrep client ipc:///tmp/reqrep_test
1712848018 - CLIENT: SENDING DATE REQUEST
1712848018 - CLIENT: WAITING FOR REPLY
1712848018 - CLIENT: CLOSING SOCKET
nng_recv error - Object closed
...
...
1712848018 - CLIENT: SENDING DATE REQUEST
1712848018 - CLIENT: WAITING FOR REPLY
1712848018 - CLIENT: CLOSING SOCKET
nng_recv error - Object closed
1712848018 - CLIENT: SENDING DATE REQUEST
1712848018 - CLIENT: WAITING FOR REPLY
1712848018 - CLIENT: CLOSING SOCKET. <--- The thread is suck in nng_close here
```

**Environment Details**

 - NNG version: 1.7.3
 - Operating system and version: MacOS Sonoma 14.4.1 (but also happens on Windows and Linux)
 - Compiler and language used: C/C++ clang
 - Shared or static library - static

**Additional context**

Here is a snapshot of the threads in the client app during the deadlock reproduction:
![Screenshot 2024-04-11 at 5 32 52 PM](https://github.com/nanomsg/nng/assets/14541274/1eea0c60-1eff-455e-b21b-457d3ae4b02b)

I recall that the initial suspect was an application callback which is not done:

> Being blocked on nni_aio_wait (as above) and especially task_wait is in indication that your application has a callback executing that is not completing.

So in this example code there is no application callback at all and only blocking APIs are called.
The expected result (as I understand) is that `nng_recv` will always return ECLOSED.

**Comment by mikisch81** [comment]
@gdamore 
I updated the modified `reqrep` example code which reproduce the issue: https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3

**Comment by leowang552** [comment]
Environment Details: 
  git hash SHA-1: 2b967f48a669f7199c40bc730f800e53d6d4fb72
  vs 2022  
  SDK 10.0.22621.0
Reproduce: 
  aio_test.c
  void test_traffic(void)
{
 	nng_socket s1;
	nng_socket s2;
	nng_aio *  tx_aio;
	nng_aio *  rx_aio;
	int        tx_done = 0;
	int        rx_done = 0;
	nng_msg *  m;
	char *     addr = "inproc://traffic";

	NUTS_PASS(nng_pair1_open(&s1));
      	NUTS_PASS(nng_listen(s1, addr, NULL, 0));
      NUTS_PASS(nng_close(s1));   // freeze 
  
}

**Comment by KnappSas** [comment]
I forgot to mention that I recently switched from nanomsg to nng and I'm using the compatibility layer

**Comment by leowang552** [comment]
> Does this occur when using synchronous APIs or only when using aio? If your callbacks are hanging that could cause this behavior.
> 
> If you have callback functions on the socket can you share them?

aio_test.c  path   nng/src/core/aio_test.c

**Comment by KnappSas** [comment]
I'm also seeing this running the tests in pair0_test.c

**Comment by gdamore** [comment]
Wait these are occurring with the included test suites not your own code???

Can I get details about the system you are using including OS, cpu, and compiler?   Is this running in a cloud or virtual environment?

**Comment by gdamore** [comment]
Ok I see the details there. Will update as soon as I have a diagnosis.

**Comment by leowang552** [comment]
I have fixed the bug, source code
```
int
nni_reap_sys_init(void)
{
	int rv;

	reap_exit = false;  // add the line
	// If this fails, we don't fail init, instead we will try to
	// start up at reap time.
	if ((rv = nni_thr_init(&reap_thr, reap_worker, NULL)) != 0) {
		return (rv);
	}
	nni_thr_run(&reap_thr);
	return (0);
}

```

**Comment by KnappSas** [comment]
I can confirm that the above from leowang fixes the problem I see in the test suit. Of course I cannot tell yet if it fixes the issue within our system too.

**Comment by mikisch81** [comment]
@gdamore can you please address the original issue I opened? Everything is detailed there, I don't want to lose context due to other examples from other users which should be in different issues.
I use sync APIs, it happens on multiple platforms and I gave a link to a simple example which easily reproduces this issue.

**Comment by leowang552** [comment]
@mikisch81  can you share your code? I tried [the code ](https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3) , but not reproduce.

**Comment by alzix** [comment]
the issue reproduces fairly easily on both macOS and Windows based on the modified [demo/reqrep](https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3) example

**Comment by leowang552** [comment]
Please post your modified code

**Comment by alzix** [comment]
```
diff --git a/demo/reqrep/reqrep.c b/demo/reqrep/reqrep.c
index 62ff29a4..3f04950f 100644
--- a/demo/reqrep/reqrep.c
+++ b/demo/reqrep/reqrep.c
@@ -72,6 +72,7 @@ showdate(time_t now)
 int
 server(const char *url)
 {
+	for (;;) {
 	nng_socket sock;
 	nng_listener listener;
 	int        rv;
@@ -100,7 +101,6 @@ server(const char *url)
 	}
 	nng_listener_start(listener, 0);
 
-	for (;;) {
 		char *   buf = NULL;
 		size_t   sz;
 		uint64_t val;
@@ -127,10 +127,22 @@ server(const char *url)
 			if (rv != 0) {
 				fatal("nng_send", rv);
 			}
+			rv  = nng_listener_close(listener);
+			if (rv != 0) {
+				fatal("nng_listener_close", rv);
+			}
+			rv  = nng_close(sock);
+			if (rv != 0) {
+				fatal("nng_close", rv);
+			}
 			continue;
 		}
 		// Unrecognized command, so toss the buffer.
 		nng_free(buf, sz);
+		rv  = nng_close(sock);
+		if (rv != 0) {
+			fatal("nng_close", rv);
+		}
 	}
 }
 
@@ -189,11 +201,11 @@ client(const char *url)
 		} else {
 			printf("CLIENT: GOT WRONG SIZE!\n");
 		}
-		nng_msleep(sleep);
-		sleep++;
-		if (sleep == 4) {
-			sleep = 4000;
-		}
+		//		nng_msleep(sleep);
+		//		sleep++;
+		//		if (sleep == 4) {
+		//			sleep = 4000;
+		//		}
 	}
 
 	// This assumes that buf is ASCIIZ (zero terminated).
```

**Comment by alzix** [comment]
the changes to make the issue more easily reproducible:
1. tweak the client code to bombard the server with requests without sleep 
2. tweak the server code to close the socket after each reply

when the server closes the socket while there is a busy receive aio task - the system goes into deadlock
based on my observations the same issue occurs when the client is in blocking `nng_recv` and the server responds just before client socket is closed.

**Comment by leowang552** [comment]
> ```
> diff --git a/demo/reqrep/reqrep.c b/demo/reqrep/reqrep.c
> index 62ff29a4..a9ab2cd8 100644
> --- a/demo/reqrep/reqrep.c
> +++ b/demo/reqrep/reqrep.c
> @@ -72,6 +72,7 @@ showdate(time_t now)
>  int
>  server(const char *url)
>  {
> +        for (;;) {
>  	nng_socket sock;
>  	nng_listener listener;
>  	int        rv;
> @@ -100,7 +101,6 @@ server(const char *url)
>  	}
>  	nng_listener_start(listener, 0);
>  
> -	for (;;) {
>  		char *   buf = NULL;
>  		size_t   sz;
>  		uint64_t val;
> @@ -127,10 +127,22 @@ server(const char *url)
>  			if (rv != 0) {
>  				fatal("nng_send", rv);
>  			}
> +                        rv  = nng_listener_close(listener);
> +                        if (rv != 0) {
> +                                fatal("nng_listener_close", rv);
> +                        }
> +                        rv  = nng_close(sock);
> +                        if (rv != 0) {
> +                                fatal("nng_close", rv);
> +                        }
>  			continue;
>  		}
>  		// Unrecognized command, so toss the buffer.
>  		nng_free(buf, sz);
> +                rv  = nng_close(sock);
> +                if (rv != 0) {
> +                        fatal("nng_close", rv);
> +                }
>  	}
>  }
>  
> @@ -189,11 +201,11 @@ client(const char *url)
>  		} else {
>  			printf("CLIENT: GOT WRONG SIZE!\n");
>  		}
> -		nng_msleep(sleep);
> -		sleep++;
> -		if (sleep == 4) {
> -			sleep = 4000;
> -		}
> +//		nng_msleep(sleep);
> +//		sleep++;
> +//		if (sleep == 4) {
> +//			sleep = 4000;
> +//		}
>  	}
>  
>  	// This assumes that buf is ASCIIZ (zero terminated).
> @@ -213,4 +225,4 @@ main(const int argc, const char **argv)
>  
>  	fprintf(stderr, "Usage: reqrep %s|%s <URL> ...\n", CLIENT, SERVER);
>  	return (1);
> -}
> +}
> \ No newline at end of file
> ```

I tested the modified code. 
Server side print nng_recv: Object closed & quit. Client is waiting receiving. 
Client print RECEIVED DATA if restart the server.
![image](https://github.com/nanomsg/nng/assets/29005122/df6900eb-7648-41d8-a129-47201f52fafc)

**Comment by alzix** [comment]
the server never exits in normal flow - in my case it hangs after 20-30 iterations.
what version/hash of `nng` are you using?

**Comment by alzix** [comment]
consider saving the diff to a text file and applying it as a git diff - or just take the whole `reqrep.c` from the [ghist](https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3)

**Comment by leowang552** [comment]
Sorry! Please forgive me for ignoring the for loop in the diff.
Environment:
git hash SHA-1: "1941fdae7eca1b92a957dfffa6a1ec7ef549dc60"

vs 2022
SDK 10.0.22621.0
win 10 64bit

result after running 10 minutes
![image](https://github.com/nanomsg/nng/assets/29005122/bce6c9da-42a2-46f5-9236-491bae2666f6)

my code file is from https://gist.github.com/mikisch81/428c4ad87afcc1c8881b282cd5e80eb3

**Comment by alzix** [comment]
IIUC the issue is caused by the fact that the `aio` we are waiting on has no cancel function.
![image](https://github.com/nanomsg/nng/assets/1101204/9e1cdd87-ad83-4ea8-9938-d4efe2ef739c)

so the implementation of `nni_aio_stop` does not actually do anything and keeps waiting forever.
```
void
nni_aio_stop(nni_aio *aio)
{
	if (aio != NULL) {
		nni_aio_cancel_fn fn;
		void             *arg;
		nni_aio_expire_q *eq = aio->a_expire_q;

		nni_mtx_lock(&eq->eq_mtx);
		nni_aio_expire_rm(aio);
		fn                = aio->a_cancel_fn;
		arg               = aio->a_cancel_arg;
		aio->a_cancel_fn  = NULL;
		aio->a_cancel_arg = NULL;
		aio->a_stop       = true;
		nni_mtx_unlock(&eq->eq_mtx);

		if (fn != NULL) {
			fn(aio, arg, NNG_ECANCELED);
		}

		nni_aio_wait(aio);
	}
} 
```

**Comment by alzix** [comment]
a potential fix is proposed in #1824 - the modified demo runs for quite a long time already on my end
